### ISSUE TITLE
Unified directory structure for generated framework.

### ISSUE THREAD

**Issue by rickteuthof** [leegbestand/coconut#10]
For the directory structure for the resulting compiler I had the following structure in mind:

```
framework
| - include
| | - core
| | - generated
| - source
| | - core
| | - generated
| - main.c
```
In this structure, "core" directories contain pre-written files containing functions that are not reliant on the dsl specification written by the user, and are therefore not necessary to be generated. The "generated" folder is self-explanatory. 

The advantage of this directory structure is that pre-written code is clearly distinguished from the generated code. Even better would be a seperate "user" directory in which the user should write their pass and/or traversal files.

With this in mind, a drawback is that we would need to have different subdirectories in the "core" directories for different backends. Right now it also makes working with the "copra" core very difficult.

What do you have in mind as a solution for this issue @leegbestand ?

**Comment by leegbestand** [comment]
My solution was to place everything that is generated in the users build directory as they are build artifacts. For copra, just include cocnut.cmake and use the target function which auto adds copra as a build dependency to the target. One thing I have not figures out well is how to have the DSL file as a cnofigure dependency for cmake. The advantage of generating at configure time is that the generated files become visible for editors to explore. 

For the different backends, we could just have copra as a central place and use #IFDEF for different backends at compile time, we can generate a #define via the ccn_enables header. Or, we can do it via a CMAKE flag and then auto add the right backend and have subdirs in copra for the backend files.

**Comment by rickteuthof** [comment]
Should it be possible for the core files in copra to be dependant on the generated files? I have multiple core files that depend on generated code in order to compile successfully. If we place these core files in copra, then copra and generated should be mutually dependant on eachother, and i'm not sure if that's even possible.

The directory structure i'm trying to achieve is currently as follows:
```
coconut
| - copra/
| | - include/
| | - source.c
| | - ...
| - generated/
| | - include/
| | - source.c
| | - ...
```
(Maybe also give the source files their own folder)

What are your thoughts on this @leegbestand ?

**Comment by leegbestand** [comment]
Copra and the generated files get compiled at the same time, so they can be dependant on each other. Copra just needs access to the header files or externs, same with the generated code. 

What is the location of coconut? The generated directory should be in the your_compiler/build directory.

coconut.cmake already defines this, here:
file(GLOB COCONUT_SOURCES ${CMAKE_BINARY_DIR}/generated/source/*.c ${COCONUT_ROOT_DIR}/copra/*.c)
    target_sources("${TARGET}" PRIVATE "${COCONUT_SOURCES}")
    target_include_directories("${TARGET}" PRIVATE ${CMAKE_BINARY_DIR}/generated/include/ ${COCONUT_ROOT_DIR}/copra/)
    target_link_libraries("${TARGET}" PRIVATE coconut::palm)

So, generate the files with cocogen, load all copra and generated files in sources and compile them at the same time. Can not do add_subdirectory(copra), because indeed copra depends on the generated stuff, therefore the GLOB(). 

From what I can see in the build stuff, Cmake automatically tracks the absolute path as a dependency in globbing. So, you get a <path_to_copra>/*.o files in your build dir. 

The current coconut.cmake should work in most parts, though. Is there some case not working? I know it doesnt automatically rebuild, but I need to figure that out. 

My final goal is to have the user only do something like:
coconut_generate(my_target, my_dsl)
and that does everything, and possible a FLAGS or set(COCONUT_FLAGS) option to enable things like consistency checking.

**Comment by leegbestand** [comment]
The current setup in new_dir_structure does not allow for parallel builds? Also, generated code and copra code should get the same flags as user code, otherwise debugging becomes annoying. The current setup indeed does not allow mutual dependency, you need to compile them at the same time for that.

**Comment by leegbestand** [comment]
Tonight, I will make a setup with your backend in an example, if required.

**Comment by rickteuthof** [comment]
The current structure in new_dir_structure should compile correctly, both copra and the generated files. It should also allow for mutual dependency now, so copra and generated can both be compiled independently, if required.

There's still a few minor issues concerning the naming conventions (lowercase/uppercase) but that's not much of an issue right now.

**Comment by leegbestand** [comment]
Ok, nice. But in the new_dir_structure, how can compiler flags be shared between the two and the user files(Not sure if this is required)?  Can we test multiple projects in parallel? How would the user do a full clean, like: rm -rf build/ ?

**Comment by rickteuthof** [comment]
That's something i'll have to look into later. We should be able to test multiple projects as long as they have correct CMakeLists.txt files

**Comment by leegbestand** [comment]
But generated files are stored in the coconut repo and not in the project's repo, correct?

**Comment by rickteuthof** [comment]
As of now, yes, I initially thought we should have a "framework" folder containing copra and generated code, and then have the user write their own code in a separate user project folder. Then from that folder we could link the user files with the files in the framework somehow.

**Comment by leegbestand** [comment]
So, why not keep the structure in coconut.cmake?

**Comment by rickteuthof** [comment]
I though it's good practice to have a src and an include folder, which is why i changed things around this way.

**Comment by leegbestand** [comment]
Yeah, that is fine. I was talking about placing the generated files in the coconut repo. Why not make them part of the users build directory?

**Comment by rickteuthof** [comment]
But then shouldn't copra know where to find those user generated files?

**Comment by leegbestand** [comment]
Files in copra are basically static generated files, as in, we could also generate them as part of a cocogen run, but they always result in the same file. So, it is easier to take it out. Therefore, you can handle copra files the same way as generated files and just glob them into the generated sources.

**Comment by leegbestand** [comment]
Also, merge backends branch can not compile coconut, because generated src/ and include/ do not exists. generated and copra can not be a subdir of coconut.

**Comment by rickteuthof** [comment]
Closed as this is solved in issue #12
